content/PreparedStatement/PreparedStatement.q.md

```diff
@@ -0,0 +1,1707 @@
+        echo $e->getMessage();
+    }
+?>
+```
+
+
+:::
+
+/// type=REPL, readonly=true, filename=[connection.php,student.php]
+
+```php
+// connection.php
+<?php
+    $host = 'localhost';
+    $db = 'LibraryDB';
+    $port = '5432';
+    $username = 'postgres';
+    $password = 'Admin01';
+
+    try {
+        $conn = new PDO("pgsql:host=$host;port=$port;dbname=$db;user=$username;password=$password");
+        if ($conn) {
+            echo "Successfully connected to the database. ";
+        }
+    } catch (Exception $e) {
+        echo "Unable to establish a connection."; 
+    }
+?>
+```
+
+```php
+// student.php
+<?php
+    require_once("connection.php");
+
+    $sql = "SELECT * FROM students WHERE gender = ?";
+
+    $pstmt = $conn->prepare();
+    $pstmt->execute(array('F'));
+    $data = $pstmt->fetchAll();
+    
+    foreach ($data as $row) {
+        echo $row['student_id'].$row['first_name'].$row['last_name'].$row['birth_date'].$row['gender']. ' ';
+    }
+?>
+```
+/// type=MS, answer=[4,5]
+
+Execute the program. What are the error messages?
+
+- Undefined variable: `pstmt` on line 7
+
+- Undefined variable: `data` on line 10
+
+- Invalid argument supplied for `foreach()` on line 10
+
+- `PDO::prepare()` expects at least `1` parameter, `0` given on line 6
+
+- Uncaught Error: Call to a member function `execute()` on boolean on line 7
+
+
+/// type=MS, answer=[2,5]
+
+Which statements correctly describe the error?
+
+- On line 4, the `students` table does not exist.
+
+- On line 6, the statement `$pstmt = $conn->prepare();` is incorrect.
+
+- In `student.php`, there is no semicolon `;` at the end of the statement on line 4.
+
+- In `student.php`, there is no close parenthesis `)` at the end of the statement on line 7.
+
+- There is no argument `$sql` specified in the statement `$pstmt = $conn->prepare();` on line 6.
+
+:::
+
+
+/// type=CR, answer=[tests/DataObjects/MissingArgumentInPrepare.php], filename=[connection.php,student.php]
+
+Correct the code so that it outputs the strings `Successfully connected to the database.` and `2SamanthaDanes1999-10-12F 4VikkiFernando2005-05-17F 6AlisaElls1999-06-30F`.
+
+```php
+// connection.php
+<?php
+    $host = 'localhost';
+    $db = 'LibraryDB';
+    $port = '5432';
+    $username = 'postgres';
+    $password = 'Admin01';
+
+    try {
+        $conn = new PDO("pgsql:host=$host;port=$port;dbname=$db;user=$username;password=$password");
+        if ($conn) {
+            echo "Successfully connected to the database. ";
+        }
+    } catch (Exception $e) {
+        echo "Unable to establish a connection."; 
+    }
+?>
+```
+
+```php
+// student.php
+<?php
+    require_once("connection.php");
+
+    $sql = "SELECT * FROM students WHERE gender = ?";
+
+    $pstmt = $conn->prepare();
+    $pstmt->execute(array('F'));
+    $data = $pstmt->fetchAll();
+    
+    foreach ($data as $row) {
+        echo $row['student_id'].$row['first_name'].$row['last_name'].$row['birth_date'].$row['gender']. ' ';
```

Comment:
I'm so sure about this. At what real world situation can we use this data?

Reply:
Okay manager, let me go ahead and make changes. Thanks.